Xamarin.PropertyEditing/ViewModels/MaterialDesignColorViewModel.cs

```diff
@@ -0,0 +1,251 @@
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Diagnostics;
+using System.Linq;
+using Xamarin.PropertyEditing.Drawing;
+using Strings = Xamarin.PropertyEditing.Properties.Resources;
+
+namespace Xamarin.PropertyEditing.ViewModels
+{
+	internal class MaterialDesignColorViewModel : NotifyingObject
+	{
+		public MaterialDesignColorViewModel (BrushPropertyViewModel parent)
+		{
+			Parent = parent ?? throw new ArgumentNullException (nameof (parent));
+			parent.PropertyChanged += Parent_PropertyChanged;
+		}
+
+		public CommonColor Color
+		{
+			get => Parent.Value is CommonSolidBrush solidBrush ? solidBrush.Color : new CommonColor (0, 0, 0);
+			private set {
+				if (!Color.Equals (value)) {
+					Parent.Value = new CommonSolidBrush (value, null, Parent.Value.Opacity);
+					OnPropertyChanged (nameof (Alpha));
+					OnPropertyChanged ();
+				}
+			}
+		}
+
+		public string ColorName
+		{
+			get {
+				if (this.colorName != null) return this.colorName;
+				if (Parent.Value is CommonSolidBrush solidBrush) {
+					return (this.colorName = Scale.Name);
+				}
+				return (this.colorName = Strings.MaterialColorGrey);
+			}
+			set {
+				// Attempt to retain normal and accent selections
+				if (NormalColor != null) {
+					var normalColorIndex = Array.IndexOf (NormalColorScale.ToArray (), NormalColor);
+					this.colorName = value;
+					CommonColor?[] newScale = NormalColorScale.ToArray();
+					if (newScale.Length > normalColorIndex) {
+						NormalColor = newScale[normalColorIndex];
+					}
+					ReplaceColor (NormalColor.Value);
+				} else if (AccentColor != null) {
+					var accentColorIndex = Array.IndexOf (AccentColorScale.ToArray (), AccentColor);
+					this.colorName = value;
+					CommonColor?[] newScale = AccentColorScale.ToArray ();
+					if (newScale.Length > accentColorIndex) {
+						AccentColor = newScale[accentColorIndex];
+					}
+					ReplaceColor (AccentColor.Value);
+				} else {
+					this.colorName = value;
+				}
+				OnPropertyChanged ();
+				OnPropertyChanged (nameof (Scale));
+				OnPropertyChanged (nameof (AccentColorScale));
+				OnPropertyChanged (nameof (NormalColorScale));
+			}
+		}
+
+		public byte Alpha
+		{
+			get => Parent.Value is CommonSolidBrush solidBrush ? solidBrush.Color.A : (byte)255;
+			set {
+				CommonColor oldColor = Color;
+				Color = new CommonColor (oldColor.R, oldColor.G, oldColor.B, value);
+			}
+		}
+
+		public MaterialColorScale Scale
+			=> MaterialPalettes
+				.OrderBy (palette => palette.Colors.Min (
+					paletteColor => CommonColor.SquaredDistance (paletteColor, Color)))
+				.First ();
+
+		public CommonColor? NormalColor
+		{
+			get {
+				if (this.normalColor.HasValue) return this.normalColor.Value;
+				MaterialColorScale scale = FindPalette (ColorName, false);
+				foreach (CommonColor normalColor in scale.Colors) {
```

Comment:
It looks like FindPalette can return null so this needs to handle that. Actually there are a lot of places where FindPalette results are accessed without a check?

Reply:
Yes, you're right.

Reply:
👍